File: dotnet/dtx/Dtx.cs
namespace Dtx;

/// <summary>
/// The header every DTX variant shares, and the numbers the variants take.
///
/// <para>doc/SPEC.md section 1: DTX, the variant, R, C, RR, the width every
/// value takes, and one zero byte, so the payload begins on a long. Every
/// field of more than one byte is most significant byte first.</para>
/// </summary>
public static class Format
{
    /// <summary>The variant byte of a table laid out row by row.</summary>
    public const int Dtx0 = 0;

    /// <summary>The variant byte of a table laid out column by column.</summary>
    public const int Dtx1 = 1;

    /// <summary>Column by column and packed.</summary>
    public const int Dtx2 = 2;

    /// <summary>The three bytes a file opens with.</summary>
    public static readonly byte[] Magic = { (byte)'D', (byte)'T', (byte)'X' };

    /// <summary>The largest ring a payload can define: N is two bytes.</summary>
    public const int MaxRing = 65535;

    /// <summary>
    /// The flags bit at payload byte 3 that marks every column was packed
    /// with copies from its own literal stream, R5.10.
    /// </summary>
    public const int CopiesFlag = 1;

    /// <summary>What a header runs to, under every variant and every C.</summary>
    public const int HeaderLength = 16;

    /// <summary>at up to the next multiple of to.</summary>
    public static int Align(int at, int to) => (at + to - 1) / to * to;

    /// <summary>The header at the start of a file.</summary>
    /// <exception cref="ArgumentException">where the file is short of a
    /// header, does not open with DTX, or breaks a bound R6 sets</exception>
    public static Header ReadHeader(byte[] file)
    {
        if (file.Length < HeaderLength)
        {
            throw new ArgumentException(
                    $"a file of {file.Length} bytes does not contain a header");
        }
        for (int i = 0; i < Magic.Length; i++)
        {
            if (file[i] != Magic[i])
            {
                throw new ArgumentException("the file does not open with DTX");
            }
        }
        int rows = GetLong(file, 4);
        int columns = GetWord(file, 8);
        int repeat = GetLong(file, 10);
        int width = file[14];
        if (rows < 1)
        {
            throw new ArgumentException($"R is 1 upward, not {rows}");
        }
        if (columns < 1 || columns > 256)
        {
            throw new ArgumentException($"C is 1 to 256, not {columns}");
        }
        if (repeat < 0 || repeat > rows)
        {
            throw new ArgumentException($"RR is 0 to R, not {repeat}");
        }
        if (width != 1 && width != 2 && width != 4)
        {
            throw new ArgumentException(
                    $"the width is 1, 2 or 4 bytes, not {width}");
        }
        return new Header(file[3], rows, columns, repeat, width);
    }

    // The four byte order helpers. A DTX file and a 68000 image are both
    // most significant byte first.

    public static int GetWord(byte[] file, int at) =>
            file[at] << 8 | file[at + 1];

    public static int GetLong(byte[] file, int at) =>
            file[at] << 24 | file[at + 1] << 16 | file[at + 2] << 8 | file[at + 3];

    public static void PutWord(byte[] file, int at, int value)
    {
        file[at] = (byte)(value >> 8);
        file[at + 1] = (byte)value;
    }

    public static void PutLong(byte[] file, int at, int value)
    {
        file[at] = (byte)(value >> 24);
        file[at + 1] = (byte)(value >> 16);
        file[at + 2] = (byte)(value >> 8);
        file[at + 3] = (byte)value;
    }
}

/// <summary>
/// What a file's header defines: the byte at offset 3, R, C, RR, and the
/// bytes every value takes.
/// </summary>
public sealed record Header(int Variant, int Rows, int Columns, int Repeat,
        int Width)
{
    /// <summary>What the header runs to, the payload's first byte.</summary>
    public int Length => Format.HeaderLength;

    /// <summary>A row's bytes: C values of W bytes.</summary>
    public int RowBytes => Columns * Width;
}

File: go/dtx/dtx.go
// Package dtx reads the header every DTX variant shares.
//
// doc/SPEC.md section 1: DTX, the variant, R, C, RR, the width every value
// takes, and one zero byte, so the payload begins on a long. Every field of
// more than one byte is most significant byte first.
package dtx

import "fmt"

// The variant byte at offset 3.
const (
	DTX0 = 0 // laid out row by row
	DTX1 = 1 // laid out column by column
	DTX2 = 2 // laid out column by column and packed
)

// Magic is the three bytes a file opens with.
var Magic = []byte{'D', 'T', 'X'}

// HeaderLength is what a header runs to, under every variant and every C.
const HeaderLength = 16

// MaxRows is the largest R a header defines, R6.1: the field is four bytes,
// and a value past what a signed long gives is out of bounds, so the error
// gives the count the header defines rather than a negative one. RR is 0 to
// R, so the one bound covers both.
const MaxRows = 2147483647

// Align gives at up to the next multiple of to.
func Align(at, to int) int {
	return (at + to - 1) / to * to
}

// Header gives what a file's header defines.
type Header struct {
	Variant int
	Rows    int // R
	Columns int // C
	Repeat  int // RR
	Width   int // W, the bytes every value takes
}

// Length gives what the header runs to, the payload's first byte.
func (h Header) Length() int {
	return HeaderLength
}

// RowBytes gives a row's bytes: C values of W bytes.
func (h Header) RowBytes() int {
	return h.Columns * h.Width
}

// ReadHeader gives the header at the start of file, or an error where the
// file is short of one, does not open with DTX, or breaks a bound R6 sets.
func ReadHeader(file []byte) (Header, error) {
	if len(file) < HeaderLength {
		return Header{}, fmt.Errorf(
			"a file of %d bytes does not contain a header", len(file))
	}
	for i, b := range Magic {
		if file[i] != b {
			return Header{}, fmt.Errorf("the file does not open with DTX")
		}
	}
	rows := GetLong(file, 4)
	columns := GetWord(file, 8)
	repeat := GetLong(file, 10)
	width := int(file[14])
	if rows < 1 || rows > MaxRows {
		return Header{}, fmt.Errorf("R is 1 to %d, not %d", MaxRows, rows)
	}
	if columns < 1 || columns > 256 {
		return Header{}, fmt.Errorf("C is 1 to 256, not %d", columns)
	}
	if repeat < 0 || repeat > rows {
		return Header{}, fmt.Errorf("RR is 0 to R, not %d", repeat)
	}
	if width != 1 && width != 2 && width != 4 {
		return Header{}, fmt.Errorf(
			"the width is 1, 2 or 4 bytes, not %d", width)
	}
	return Header{int(file[3]), rows, columns, repeat, width}, nil
}

// The four byte order helpers. A DTX file and a 68000 image are both most
// significant byte first.

// GetWord reads two bytes at at.
func GetWord(in []byte, at int) int {
	return int(in[at])<<8 | int(in[at+1])
}

// GetLong reads four bytes at at.
func GetLong(in []byte, at int) int {
	return int(in[at])<<24 | int(in[at+1])<<16 | int(in[at+2])<<8 | int(in[at+3])
}

// PutWord writes two bytes at at.
func PutWord(out []byte, at, value int) {
	out[at] = byte(value >> 8)
	out[at+1] = byte(value)
}

// PutLong writes four bytes at at.
func PutLong(out []byte, at, value int) {
	out[at] = byte(value >> 24)
	out[at+1] = byte(value >> 16)
	out[at+2] = byte(value >> 8)
	out[at+3] = byte(value)
}
